plugin/src/main/java/org/opensearch/ml/model/MLModelCacheHelper.java

```diff
@@ -75,6 +79,165 @@ public synchronized void setModelState(String modelId, MLModelState state) {
+    }
+
+    /**
+     * Remove the rate limiter from cache to disable model level throttling
+     * @param modelId model id
+     */
+    public synchronized void removeModelRateLimiter(String modelId) {
+        log.debug("Removing the rate limiter for Model {}", modelId);
+        getExistingModelCache(modelId).setModelRateLimiter(null);
+    }
+
+    /**
+     * Set the user rate limiter map for a single user to enable user level throttling.
+     *
+     * @param modelId model id
+     * @param user user
+     * @param rateLimiter rate limiter
+     */
+    public synchronized void setUserRateLimiterMap(String modelId, String user, TokenBucket rateLimiter) {
+        log.debug("Setting the user level rate limiter for Model {}", modelId);
+        Map<String, TokenBucket> userRateLimiterMap = new HashMap<>() {
+            {
+                put(user, rateLimiter);
+            }
+        };
+        getExistingModelCache(modelId).setUserRateLimiterMap(userRateLimiterMap);
+    }
+
+    /**
+     * Set the user rate limiter map to enable user level throttling.
+     *
+     * @param modelId model id
+     * @param userRateLimiterMap a map with user's name and its corresponding rate limiter
+     */
+    public synchronized void setUserRateLimiterMap(String modelId, Map<String, TokenBucket> userRateLimiterMap) {
+        log.debug("Setting the user level rate limiter for Model {}", modelId);
+        getExistingModelCache(modelId).setUserRateLimiterMap(userRateLimiterMap);
+    }
+
+    /**
+     * Update the user rate limiter map with the user rate limiter map.
+     * If the user rate limiter map doesn't exist for the model, consider calling setUserRateLimiterMap instead.
+     *
+     * @param modelId model id
+     * @param updateUserRateLimiterMap a map with user's name and its corresponding rate limiter
+     */
+    public synchronized void updateUserRateLimiterMap(String modelId, Map<String, TokenBucket> updateUserRateLimiterMap) {
+        log.debug("Updating the user level rate limiter for Model {}", modelId);
+        Map<String, TokenBucket> userRateLimiterMap = getExistingModelCache(modelId).getUserRateLimiterMap();
+        if (userRateLimiterMap != null) {
+            userRateLimiterMap.putAll(updateUserRateLimiterMap);
+        } else {
+            throw new OpenSearchStatusException(
+                "Model controller doesn't exist for the model. Consider calling create model controller api instead. Model ID: " + modelId,
+                RestStatus.CONFLICT
+            );
+        }
+    }
+
+    /**
+     * Update the user rate limiter map for a single user.
+     * If the user rate limiter map doesn't exist for the model, consider calling setUserRateLimiterMap instead.
+     *
+     * @param modelId model id
+     * @param user user
+     * @param rateLimiter rate limiter
+     */
+    public synchronized void updateUserRateLimiterMap(String modelId, String user, TokenBucket rateLimiter) {
```

Comment:
A general question for all the update/get in the MLModelCache, do we need to check the data consistency with the indices before returning the results? Or they are all taken care in the transport layer actions?

Reply:
We will only update the cache once we confirm the index has been updated. That's how we control the consistency.